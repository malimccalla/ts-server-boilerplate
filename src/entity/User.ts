import * as bcrypt from 'bcryptjs';
import {
  BaseEntity,
  BeforeInsert,
  Column,
  Entity,
  PrimaryGeneratedColumn,
} from 'typeorm';

@Entity('users')
export class User extends BaseEntity {
  @PrimaryGeneratedColumn('uuid') id: string;

  @Column('varchar', { length: 255 })
  email: string;

  @Column('text') password: string;

  @Column('boolean', { default: false })
  confirmed: boolean;

  @Column('boolean', { default: false })
  locked: boolean;

  @BeforeInsert()
  async hashPassword() {
    const password = await bcrypt.hash(this.password, 10);
    this.password = password;
  }
}
