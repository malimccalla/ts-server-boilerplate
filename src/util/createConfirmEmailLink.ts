import { Redis } from 'ioredis';
import { v4 } from 'uuid';

export const createConfirmEmailLink = async (
  url: string,
  userId: string,
  redis: Redis
) => {
  const oneDay = 60 * 60 * 24;
  const id = v4();
  await redis.set(id, userId, 'ex', oneDay);

  return `${url}/confirm/${id}`;
};
