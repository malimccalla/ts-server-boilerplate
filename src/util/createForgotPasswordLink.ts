import { Redis } from 'ioredis';
import { v4 } from 'uuid';

import { forgotPasswordPrefix } from '../constants';

export const createForgotPasswordLink = async (
  url: string,
  userId: string,
  redis: Redis
) => {
  const twentyMins = 60 * 20;
  const key = v4();
  await redis.set(`${forgotPasswordPrefix}${key}`, userId, 'ex', twentyMins);

  return { key, link: `${url}/change-password/${key}` };
};
