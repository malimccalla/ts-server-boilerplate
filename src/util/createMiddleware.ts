import { GraphQLMiddlewareFunc, Resolver } from '../types';

export const createMiddleware = (
  middlewareFunc: GraphQLMiddlewareFunc,
  resolverFunc: Resolver
) => (parent: any, args: any, context: any, info: any) => {
  console.log('CREATE MIDDLEWARE');

  return middlewareFunc(resolverFunc, parent, args, context, info);
};
