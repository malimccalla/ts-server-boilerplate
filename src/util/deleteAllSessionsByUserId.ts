import { Redis } from 'ioredis';

import { redisSessionPrefix, userSessionIdPrefix } from '../constants';
import { Session } from '../types';

export const deleteAllSessionsByUserId = async (
  userId: string,
  redis: Redis,
  session: Session
) => {
  const sessionIds = await redis.lrange(`${userSessionIdPrefix}${userId}`, 0, -1);

  const rPipeline = redis.multi();

  sessionIds.forEach((key: string) => {
    rPipeline.del(`${redisSessionPrefix}${key}`);
  });

  await rPipeline.exec(err => {
    if (err) {
      console.log(err);
    }
  });

  session.destroy(err => {
    if (err) {
      console.log(err);
    }
  });
};
