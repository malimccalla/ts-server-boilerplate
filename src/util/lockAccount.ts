import { Redis } from '../../node_modules/@types/ioredis';
import { User } from '../entity/User';
import { Session } from '../types';
import { deleteAllSessionsByUserId } from './deleteAllSessionsByUserId';

const lockAccount = async (userId: string, redis: Redis, session: Session) => {
  await User.update({ id: userId }, { locked: true });
  await deleteAllSessionsByUserId(userId, redis, session);
};

export default lockAccount;
