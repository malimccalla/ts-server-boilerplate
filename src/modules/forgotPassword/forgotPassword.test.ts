import * as faker from 'faker';
import * as Redis from 'ioredis';
import { Connection } from 'typeorm';

import { User } from '../../entity/User';
import { createTestConn } from '../../test/createTestConn';
import { TestClient } from '../../test/TestClient';
import { createForgotPasswordLink } from '../../util/createForgotPasswordLink';

faker.seed(Date.now() + Math.random());

describe('Forgot password', () => {
  const email = faker.internet.email();
  const oldPassword = faker.internet.password();
  const newPassword = 'whenDovesCry';
  const redis = new Redis();
  let conn: Connection;
  let userId: string;

  beforeEach(async () => {
    conn = await createTestConn();

    const user = await User.create({
      email,
      password: oldPassword,
      confirmed: true,
    }).save();
    userId = user.id;
  });

  afterEach(async () => {
    await conn.close();
  });

  test('Should set a new password', async () => {
    const client = new TestClient();
    const { key } = await createForgotPasswordLink('', userId, redis);

    const res = await client.forgotPasswordChange({ newPassword, key });

    expect(res.forgotPasswordChange.ok).toBe(true);
  });

  test('Should not set a new password with wrong key', async () => {
    const client = new TestClient();

    const res = await client.forgotPasswordChange({
      newPassword,
      key: 'purpleRain',
    });

    expect(res.forgotPasswordChange.ok).toBe(false);
    expect(res.forgotPasswordChange).toMatchSnapshot();
  });

  test('Should be able to login with the new password', async () => {
    const client = new TestClient();
    const { key } = await createForgotPasswordLink('', userId, redis);

    await client.forgotPasswordChange({ newPassword, key });
    const { data } = await client.login(email, newPassword);

    expect(data.login.ok).toBe(true);
    expect(data.login.user!.email).toEqual(email);
  });

  test('Should not be able to login with the old password', async () => {
    const client = new TestClient();
    const { key } = await createForgotPasswordLink('', userId, redis);

    await client.forgotPasswordChange({ newPassword, key });
    const { data } = await client.login(email, oldPassword);

    expect(data.login.ok).toBe(false);
    expect(data.login.user).toBeNull();
  });

  test('Should hash the new password', async () => {
    const client = new TestClient();
    const { key } = await createForgotPasswordLink('', userId, redis);

    await client.forgotPasswordChange({ newPassword, key });

    const user = await User.findOne({ where: { id: userId } });

    expect(user!.password).not.toEqual(newPassword);
  });

  test('Should not be able to use the same key twice', async () => {
    const client = new TestClient();
    const { key } = await createForgotPasswordLink('', userId, redis);

    await client.forgotPasswordChange({ newPassword, key });
    const res = await client.forgotPasswordChange({
      newPassword: 'sometimesItSnowsInApril',
      key,
    });

    expect(res.forgotPasswordChange.ok).toBe(false);
  });
});
