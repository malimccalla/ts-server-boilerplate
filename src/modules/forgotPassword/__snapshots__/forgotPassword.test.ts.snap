// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forgot password Should not set a new password with wrong key 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Expired key",
      "path": "key",
    },
  ],
  "ok": false,
}
`;
